src/rpc/misc.cpp

```diff
@@ -974,16 +971,21 @@ static RPCHelpMan getaddressbalance()
         throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, "Invalid address");
     }
 
-    std::vector<std::pair<CAddressIndexKey, CAmount> > addressIndex;
+    std::vector<CAddressIndexEntry> addressIndex;
+
+    ChainstateManager& chainman = EnsureAnyChainman(request.context);
 
-    for (const auto& address : addresses) {
-        if (!GetAddressIndex(address.first, address.second, addressIndex)) {
-            throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, "No information available for address");
+    int nHeight;
+    {
+        LOCK(::cs_main);
+        for (const auto& address : addresses) {
+            if (!GetAddressIndex(*pblocktree, address.first, address.second, addressIndex)) {
+                throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, "No information available for address");
+            }
         }
+        nHeight = chainman.ActiveChain().Height();
     }
```

Comment:
```suggestion
    int nHeight = [&]() {
        LOCK(::cs_main);
        for (const auto& address : addresses) {
            if (!GetAddressIndex(*pblocktree, address.first, address.second, addressIndex)) {
                throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, "No information available for address");
            }
        }
        return chainman.ActiveChain().Height();
    }();
```

Reply:
Fetching `nHeight` and calling `GetAddressIndex` are two independent actions that happen to access variables protected by the same `RecursiveMutex`, for that reason they were placed within the same scope. The proposed refactor makes it appears as if running `GetAddressIndex` is part of the process for fetching `nHeight`, when it's not.

Reply:
disagree but nit